Firestore/core/src/firebase/firestore/local/leveldb_opener.cc

```diff
@@ -0,0 +1,250 @@
+
+#include <string>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/core/database_info.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_persistence.h"
+#include "Firestore/core/src/firebase/firestore/local/local_serializer.h"
+#include "Firestore/core/src/firebase/firestore/remote/serializer.h"
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+#include "Firestore/core/src/firebase/firestore/util/log.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+#include "absl/strings/match.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+namespace {
+
+using core::DatabaseInfo;
+using remote::Serializer;
+using util::Filesystem;
+using util::Path;
+using util::Status;
+using util::StatusOr;
+using util::StringFormat;
+
+constexpr const char* kReservedPathComponent = "firestore";
+
+Status FromCause(const std::string& message, const Status& cause) {
+  if (cause.ok()) return cause;
+
+  return Status(cause.code(), message).CausedBy(cause);
+}
+
+}  // namespace
+
+LevelDbOpener::LevelDbOpener(DatabaseInfo database_info, Filesystem* fs)
+    : database_info_(std::move(database_info)),
+      fs_(fs ? fs : Filesystem::Default()) {
+}
+
+LevelDbOpener::LevelDbOpener(DatabaseInfo database_info, Path app_data_dir)
+    : database_info_(std::move(database_info)),
+      app_data_dir_(std::move(app_data_dir)),
+      fs_(Filesystem::Default()) {
+}
+
+util::StatusOr<std::unique_ptr<LevelDbPersistence>> LevelDbOpener::Create(
+    const LruParams& lru_params) {
+  auto maybe_dir = PrepareDataDir();
+  if (!maybe_dir.ok()) return maybe_dir.status();
+  Path db_data_dir = maybe_dir.ValueOrDie();
+
+  LOG_DEBUG("Using %s for LevelDB storage", db_data_dir.ToUtf8String());
+
+  Serializer remote_serializer(database_info_.database_id());
+  LocalSerializer local_serializer(std::move(remote_serializer));
+
+  return LevelDbPersistence::Create(db_data_dir, std::move(local_serializer),
+                                    lru_params);
+}
+
+StatusOr<Path> LevelDbOpener::LevelDbDataDir() {
+  StatusOr<Path> maybe_dir = FirestoreAppDataDir();
+  if (!maybe_dir.ok()) return maybe_dir;
+  return StorageDir(maybe_dir.ValueOrDie());
+}
+
+StatusOr<Path> LevelDbOpener::PrepareDataDir() {
+  StatusOr<Path> maybe_dir = LevelDbDataDir();
+  if (!maybe_dir.ok()) return maybe_dir;
+  Path db_data_dir = std::move(maybe_dir).ValueOrDie();
+
+  // Check for the preferred location. If it exists, we're done.
+  Status dir_status = fs_->IsDirectory(db_data_dir);
+  if (dir_status.ok()) {
+    return db_data_dir;
+  } else if (dir_status.code() != Error::NotFound) {
+    return dir_status;
+  }
+
+  // The preferred dir doesn't exist so check for the legacy location. If it
+  // exists, migrate.
+  maybe_dir = FirestoreLegacyAppDataDir();
```

Comment:
I had trouble making this work. The issue is that when the directory is not found I need both the Status and the Path in the return value. I tried to make this work by making a `Path` out-parameter but the interface is just clunky. I ended up returning `std::tuple<Status, Path, bool>`. LMK if you think this is an improvement.

Reply:
Erg, sorry...  that didn't turn out like I was hoping / expecting at all.  It doesn't seem like an improvement. :-/  Is it easy to roll back to the old code?

FWIW it may still be flawed, but my original intention was:
1. Only use VerifyDirectoryExists() when checking for the legacy directory.  I think this removes the need to return both `exists` and the `path`...
2. VerifyDirectoryExists() would just pass-through any already-failed `StatusOr<>` object passed in (e.g. UNIMPLEMENTED) else it would return NOT_FOUND if the path doesn't exist, else it would return the path if it exists.
3. Then the code here basically becomes:
```
auto legacy_dir = VerifyDirectoryExists(FirestoreLegacyAppDataDir());
if (legacy_dir.ok()) { 
  /* migrate */ 
} else if (legacy_dir.code() != NOT_FOUND && legacy_dir.code() != UNIMPLEMENTED) {
  return legacy_dir;
} else {
  /* No legacy dir. Just create empty data dir */
}
```

But it's probably not worth spending more time on this.  Feel free to just go back to the old code and call it done.

Reply:
Reverted to the prior state + the removal of the special case for Unimplemented suggested below.

Reply:
LGTM.